components/serializers.py

```diff
@@ -159,3 +195,92 @@ class Meta:
             "catalog",
             "controls_count",
         )
+
+
+class ComponentControlSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Component
+        fields = (
+            "pk",
+            "controls",
+            "description",
+        )
+
+    def update(self, instance, validated_data):
+        # print("controlId:", validated_data.get("control_id"))
+        # if control_id := validated_data.get("control_id"):
+        #     print("controlId:", control_id)
+        print("starting json", instance.component_json)
+        if controls := validated_data.get(
+            "controls"
+        ):  # and controls not in instance.controls:
+            print("controls", controls)
+            print("existing controls", instance.controls)
+            existing_control = controls not in instance.controls
+            print("existing_control", existing_control)
+            # instance.controls.append(controls)
+            if not existing_control:
+                instance.controls = list(set(instance.controls).union(controls))
+
+            print(controls)
+            if description := validated_data.get("description"):
+                # data = collect_component_data(component_json)
+                print(description)
+                # print("existing json", instance.component_json)
+                implemented_requirement = {
+                    "uuid": str(uuid.uuid4()),
+                    "props": [
+                        {
+                            "name": "security_control_type",
+                            "uuid": str(uuid.uuid4()),
+                            "value": "Hybrid",
+                        },
+                        {"name": "provider", "uuid": str(uuid.uuid4()), "value": "No"},
+                    ],
+                    "control-id": controls[0],
+                    "description": description,
+                }
+                if existing_control:
+                    for implemented in (
+                        instance.component_json.get("component-definition")
+                        .get("components")[0]
+                        .get("control-implementations")[0]
+                        .get("implemented-requirements")
+                    ):
+                        # form_values["add"].append({"value": a.id, "label": a.title})
+                        if implemented.get("control-id") == controls[0]:
```

Comment:
@jwhite3 For some reason this only removes 1 "implemented-requirements" with a matching control-id, which we should only have 1, but when I broke the data and have a duplicate, it still leaves a duplicate :/

Reply:
Python ends the loop on successful if condition... This is fine as long as we keep data clean.